
#' Read DRAGEN summary information
#'
#'  For each sample, DRAGEN metagenomics creates a summary file (csv) 
#'  that contains counts per species that can be added at any taxonomic level 
#'  of interest. Here we read and prepare for later merging or additional analysis
#'  
#' @param file_sample Absolute path to file
#' @param nameGiven Name to use for sample identification
#' @param tax_levels Taxonomic levels to import. Default: c("Kingdom","Phylum","Class","Order","Family","Genus", "Species")
#'
#' @returns Dataframe
#' @export
read_dragen_sample_summary <- function(file_sample, nameGiven,
                                tax_levels = c("Kingdom","Phylum","Class",
                                              "Order","Family","Genus", "Species")) {
  
  print("+ Reading sample: ")
  print(paste0("++ ", nameGiven))
  ##
  sample_data <- read.csv(file_sample, header = TRUE, 
                          sep = ",", dec = ".", check.names = FALSE)
  
  print("++ Create unique ID")
  # a unique ID is generated by checking the taxonomic information for each OTU and creating a unique
  # identifier using the digest package
  library(digest)
  sample_data['id'] <- lapply(X = 1:nrow(sample_data), function(x){
    digest(as.list(sample_data[x,tax_levels]))
  }) %>% unlist
  rownames(sample_data) <- sample_data$id
  
  colnames(sample_data)[8] <- nameGiven
  print("...")
  
  sample_data
}

#' Get dragen metagenomics data for all samples
#'
#' @param sample_names Vector of sample names to use
#' @param file_paths Vector of absolute path for each sample
#'
#' @returns List of dataframes containing counts, taxonomy and all information for each sample
#' @export
get_dragenData <- function(sample_names, file_paths) {
  
  original_Data <- list()
  data_samples <- list()
  OTUs_ids <- list()
  counts_sample <- list()
  
  ## taxonomy
  taxonomy_df <- data.frame()
  
  tax_levels <- c("Kingdom","Phylum","Class",
                  "Order","Family","Genus", "Species")
  
  for (x in c(1:length(sample_names))) {
    sample_n <- sample_names[x]
    data_s <- read_dragen_sample_summary(file_sample = file_paths[x],
                                         nameGiven = sample_n)
    
    original_Data[[sample_n]] <- data_s
    data_samples[[sample_n]] <- data_s[,c('id', sample_n)]
    OTUs_ids[[sample_n]] <- rownames(data_s)
    counts_sample[[sample_n]] <- sum(data_s[sample_n])
    
    # taxonomy
    taxonomy_df <- rbind(taxonomy_df, data_s[,c('id', tax_levels)]) %>% unique()
  }
  
  ## return all information obtained
  return(
    list(
      'all_data_samples' = original_Data, 
      'data_samples' = data_samples,
      'OTUs_sample' = OTUs_ids,
      'counts_sample' = counts_sample,
      'taxonomy_df' = taxonomy_df
    ))
  
}

#' Produce OTU matrix from dragen imported data
#'
#' @param dragen_data_samples get_dragenData$data_samples result generated. A list of dataframes for eac sample
#'
#' @returns Data frame with OTUs for each sample
get_OTUs_dragen <- function(dragen_data_samples) {
  library(tidyverse)
  OTU_table <- dragen_data_samples %>% reduce(full_join, by="id")
  OTU_table[is.na(OTU_table)] = 0
  
  ## return
  OTU_table
}


#' dragen2phyloseq: import illumina DRAGEN metagenomics into phyloseq object
#'
#' @param sample_sheet Metadata information stored as a dataframe for later phyloseq. This dataframe must contain a column with sample ids and another with each file sample absolute path
#' @param col_name Column name to use for sample names. Default: id
#' @param col_filepath Column name to use for absolute file path. Default: abs_path
#'
#' @export
dragen2phyloseq <- function(sample_sheet, col_name="id", col_filepath="abs_path" ) {
  
  ##---------------------------
  ## Import dragen metagenomics for each sampl
  ##---------------------------
  dragenData <- get_dragenData(sample_names = sample_sheet[[col_name]], 
                               file_paths = sample_sheet[[col_filepath]])
  
  OTU_table <- get_OTUs_dragen(dragen_data_samples = dragenData$data_samples)
  
  ##---------------------------
  ## Create Phyloseq object
  ##---------------------------
  # Add OTU table
  rownames(OTU_table) <- OTU_table$id
  OTU_table$id <- NULL
  
  OTU <- phyloseq::otu_table(OTU_table, taxa_are_rows = TRUE)
  
  # taxonomy table
  dragenData$taxonomy_df$id <- NULL
  colnames(dragenData$taxonomy_df)[1] <- "Domain"
  TAX <- phyloseq::tax_table(object = as.matrix(dragenData$taxonomy_df))
  
  # sample sheet
  head(sample_sheet)
  sampledata <- phyloseq::sample_data(sample_sheet)
  
  # create phyloseq
  physeq <- phyloseq(OTU, TAX, sampledata)
  
  return(list('dragenData'=dragenData, 
              "physeq"=physeq))
  
}


